common/loop.c

```diff
@@ -0,0 +1,175 @@
+#include <limits.h>
+#include <string.h>
+#include <stdbool.h>
+#include <stdlib.h>
+#include <stdio.h>
+#include <poll.h>
+#include <time.h>
+#include <unistd.h>
+#include "list.h"
+#include "log.h"
+#include "loop.h"
+
+struct loop_fd_event {
+	void (*callback)(int fd, short mask, void *data);
+	void *data;
+};
+
+struct loop_timer {
+	void (*callback)(void *data);
+	void *data;
+	struct timespec expiry;
+};
+
+struct loop {
+	struct pollfd *fds;
+	int fd_length;
+	int fd_capacity;
+
+	list_t *fd_events; // struct loop_fd_event
+	list_t *timers; // struct loop_timer
+};
+
+struct loop *loop_create(void) {
+	struct loop *loop = calloc(1, sizeof(struct loop));
+	if (!loop) {
+		wlr_log(WLR_ERROR, "Unable to allocate memory for loop");
+		return NULL;
+	}
+	loop->fd_capacity = 10;
+	loop->fds = malloc(sizeof(struct pollfd) * loop->fd_capacity);
+	loop->fd_events = create_list();
+	loop->timers = create_list();
+	return loop;
+}
+
+void loop_destroy(struct loop *loop) {
+	list_foreach(loop->fd_events, free);
+	list_foreach(loop->timers, free);
+	list_free(loop->fd_events);
+	list_free(loop->timers);
+	free(loop);
```

Comment:
Is there a leak here of `loop->fds`?

Reply:
There's also a leak because elements in the lists aren't free'd.

Reply:
Not any more ;)